### PR TITLE
Gradient test based example without wrappers

### PR DESCRIPTION
**Pull request by navjotk**
This PR adds an example in the `examples` folder based on `test_gradient`, however it reproduces this test without any wrappers on the operators. This way the operators are available directly in the example code to play around with, for feature development.

This is especially helpful for #216 since this can be the basis for the example that is required to build the interface to the revolve checkpointer. 

This can serve as an example of how something complex like the gradient test can be implemented with pure Operators with (relative) ease. 

This PR also adds a utility function to the `CMemory` class (the one that wraps devito allocated memory) to do a quick `memset`. This is required to be able to reuse the same block of memory for multiple passes.

This PR includes the changes from #241 so any review should keep that in mind.

### CODE REVIEW

File: examples/acoustic/gradient_example.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+                        dtype=model.dtype)
+    u_nosave.pad_time = False
+
+    # Forward wavefield where all timesteps are stored
+    # With checkpointing this should go away <----
+    u_save = TimeData(name="u_s", shape=model.shape_domain, time_dim=nt,
+                      time_order=time_order, space_order=space_order, save=True,
+                      dtype=model.dtype)
+    u_save.pad_time = True
+
```
Comment (mloubout) [inline examples/acoustic/gradient_example.py]: As discussed with @mlange05 and in the meeting, I am not a big fan of all this initializations in the main script. Specially having objects like `Dimensions` that should not be seen by the end user.

Comment (FabioLuporini) [inline examples/acoustic/gradient_example.py]: Remind me what that ``pad_time`` does?

**Comment by FabioLuporini** [comment]
Looks good to me

File: examples/acoustic/gradient_example.py (hunk 1 above)
Comment (mloubout) [inline examples/acoustic/gradient_example.py]: Outdated reminder of the previous Operator, should be removed.

**Comment by navjotk** [comment]
- `Dimension` objects are meant to be seen by the application developer, who is the target audience for this example. They are part of devito's front-end API. This serves as an example of how to use them.

- If someone who is generally aware of how pde-constrained inversion works but not in the seismic context wants to write a solver using devito but for another pde, the examples using the seismic wrappers wouldn't be very helpful to them since they don't share the same vocabulary. This example tells them how to do the most essential things in inversion without any wrappers, which is what someone needs to see when they want to use devito for another domain for which wrappers don't exist.

File: examples/acoustic/gradient_example.py (hunk 1 above)
Comment (navjotk) [inline examples/acoustic/gradient_example.py]: Done

**Comment by FabioLuporini** [comment]
Ready to land?

**Comment by mlange05** [comment]
> If someone who is generally aware of how pde-constrained inversion works but not in the seismic context wants to write a solver using devito but for another pde, the examples using the seismic wrappers wouldn't be very helpful to them since they don't share the same vocabulary. This example tells them how to do the most essential things in inversion without any wrappers, which is what someone needs to see when they want to use devito for another domain for which wrappers don't exist.

Ok, I'm sorry to say this, but I agree with @mloubout that this increased complexity is unnecessary and I actually think that this merge does not look particularly helpful (at least to me, and I might be wrong here). My rationale is that the gradient test is possibly the most complicated one we currently have (at least from an implementation point of view), and we are using wrappers in the original test to hide some of this complexity for a reason! If the intention is to highlight the more low-level concepts of the current Devito API (dimensions, point sources, etc.) to new user not familiar with seismic inversion, this can be done with a much simpler set of examples (eg. `adjointA.py`, maybe even `adjointJ.py`, or other example) without throwing the full complexity of the gradient test at them - it might also be easier accessible if it were done in a notebook or website tutorial with full-sentence explanations and type-set equations). And without the educational argument I see very little reason to have to maintain two equivalent versions of an extremely complicated test, where one is quite challenging to read. Possibly up for discussion in the next dev call though?

I would also like to point out that we are currently preparing for a major release, so the objective should be to reduce existing complexity and enhance documentation material, not replicate and add complexity. But again, up for discussion.

File: examples/acoustic/gradient_example.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    # Gradient symbol
+    grad = DenseData(name="grad", shape=model.shape_domain, dtype=model.dtype)
+    # Reusing u_nosave from above as the adjoint wavefield since it is a temp var anyway
+    gradop = GradientOperator(model, u_nosave, grad, rec_g, u_save, time_order=time_order,
+                              spc_order=space_order, cache_blocking=cache_blocking,
+                              dse=dse, dle=dle, compiler=compiler, profile=True)
+
+    # Clear the wavefield variable to reuse it
+    # This time it represents the adjoint field
+    u_nosave.data.fill(0)
```
Comment (mlange05) [inline examples/acoustic/gradient_example.py]: I take it `.data.fill` is a custom Devito function? Why not stick to native Python instead `u_nosave.data[:] = 0.`?

File: examples/acoustic/gradient_example.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    src = SourceLike(name="src", dimensions=[time, p_src], npoint=nsrc, nt=nt,
+                     dt=dt, h=h, ndim=ndim, nbpml=nbpml, dtype=dtype,
+                     coordinates=location)
+    src.data[:] = time_series
+
+    # Create receiver symbol
+    p_rec = Dimension('p_rec', size=nrec)
+
+    # Receiver for true model
+    rec_t = SourceLike(name="rec_t", dimensions=[time, p_rec], npoint=nrec, nt=nt, dt=dt,
```
Comment (mlange05) [inline examples/acoustic/gradient_example.py]: If the objective is to not use seismic-specific utilities, should we not rebase this on #239 and use internal `PointData` instead?

File: examples/acoustic/gradient_example.py (hunk 2 above)
Comment (navjotk) [inline examples/acoustic/gradient_example.py]: It's a wrapper for [numpy.ndarray.fill](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.fill.html) and is [faster](http://stackoverflow.com/a/13052254)

**Comment by navjotk** [comment]
> this increased complexity

No complexity has increased. This is literally the simplest way for someone to write a similar test for a different equation. Can you write this test in a simpler way using the **Devito** API (wrappers are not the Devito API)? Show me. The wrappers are only for seismic. Do we not want people to try other equations? Are you saying this test is not a useful demo? This test is essential if you want to carry out an inversion problem. If you want complexity to decrease, maybe look at the API and not the example that demonstrates its use. 

> does not look particularly helpful (at least to me

That's because you don't understand the gradient test

> this can be done with a much simpler set of examples (eg. adjointA.py, maybe even adjointJ.py, or other example)

Maybe add some more examples to build up to this. That is no reason to throw away a perfectly good example. The difference between adjointJ.py and this is literally 5 lines of code. Are you arguing over 5 lines of code?

>done in a notebook or website tutorial with full-sentence explanations and type-set equations

Doesn't have to be one or another. You are free to do that if you think it is necessary. 

> enhance documentation material

In a situation of a complete lack of documentation, examples like this are the only documentation we have. If you would like to see some documentation, maybe write some?

Overall I think this is just nitpicking since this PR doesn't touch any devito code. Adding an example (where examples are seriously lacking) is a positive step. At no point have you given a single reason to NOT have this. 

You are free to junk this PR as well, if that makes you happy :-)

**Comment by FabioLuporini** [comment]
I think it's good to have a complete set of examples of increasing complexity. 

One observation though:

> The difference between adjointJ.py and this is literally 5 lines of code

Can we then avoid duplicate code (if we are, I haven't checked)?

**Comment by FabioLuporini** [comment]
@navjotk can you take a look at my question above?

**Comment by navjotk** [comment]
The "duplicated code" that you talk about is setup code. Setting the grid sizes, putting the sources and receivers. If you see the example that this PR proposes, no execution has happened until line 109. We are only setting up the environment in which the test will be conducted until then. There is no reason to hide that behind wrappers except maybe if you want to demonstrate this in a notebook where you could just import a pre-prepared test environment. For anything remotely realistic you would want to set this up manually since this is the stuff that changes from one situation to another. Hence it is actually educational to leave multiple versions of the setup code lying around in the examples so that people see how different things can be achieved. 

TLDR: What you see as duplicate code is the numerical analysis equivalent of variable definitions (`float a;` etc). You want to define the variables every time you use them. It would be counter-productive to write a wrapper that defines variables for you.

**Comment by FabioLuporini** [comment]
Right, all clear
